uefi/src/fs/file_system.rs

```diff
@@ -0,0 +1,303 @@
+    /// The path is invalid because of the underlying [`PathError`].
+    IllegalPath(PathError),
+    /// The file or directory was not found in the underlying volume.
+    FileNotFound(String),
+    /// The path is existent but does not correspond to a directory when a
+    /// directory was expected.
+    NotADirectory(String),
+    /// The path is existent but does not correspond to a file when a file was
+    /// expected.
+    NotAFile(String),
+    /// Can't delete the file.
+    CantDeleteFile(String),
+    /// Can't delete the directory.
+    CantDeleteDirectory(String),
+    /// Error writing bytes.
+    WriteFailure,
+    /// Error flushing file.
+    FlushFailure,
+    /// Error reading file.
+    ReadFailure,
+    /// Can't parse file content as UTF-8.
+    Utf8Error(FromUtf8Error),
+    /// Could not open the given path.
+    OpenError(String),
+}
+
+#[cfg(feature = "unstable")]
+impl core::error::Error for FileSystemError {}
+
+impl From<PathError> for FileSystemError {
+    fn from(err: PathError) -> Self {
+        Self::IllegalPath(err)
+    }
+}
+
+/// Return type for public [`FileSystem`] operations.
+pub type FileSystemResult<T> = Result<T, FileSystemError>;
+
+/// High-level file-system abstraction for UEFI volumes with an API that is
+/// close to `std::fs`. It acts as convenient accessor around the
+/// [`SimpleFileSystemProtocol`].
+pub struct FileSystem<'a>(ScopedProtocol<'a, SimpleFileSystemProtocol>);
+
+impl<'a> FileSystem<'a> {
+    /// Constructor.
+    #[must_use]
+    pub fn new(proto: ScopedProtocol<'a, SimpleFileSystemProtocol>) -> Self {
+        Self(proto)
+    }
+
+    /// Tests if the underlying file exists. If this returns `Ok`, the file
+    /// exists.
+    pub fn try_exists(&mut self, path: impl AsRef<Path>) -> FileSystemResult<()> {
+        self.metadata(path).map(|_| ())
+    }
+
+    /// Copies the contents of one file to another. Creates the destination file
+    /// if it doesn't exist and overwrites any content, if it exists.
+    pub fn copy(
+        &mut self,
+        src_path: impl AsRef<Path>,
+        dest_path: impl AsRef<Path>,
+    ) -> FileSystemResult<()> {
+        let read = self.read(src_path)?;
+        self.write(dest_path, read)
+    }
+
+    /// Creates a new, empty directory at the provided path
+    pub fn create_dir(&mut self, path: impl AsRef<Path>) -> FileSystemResult<()> {
+        let path = path.as_ref();
+        self.open(path, UefiFileMode::CreateReadWrite, true)
+            .map(|_| ())
+            .map_err(|err| {
+                log::debug!("failed to fetch file info: {err:#?}");
+                FileSystemError::OpenError(path.to_string())
+            })
+    }
+
+    /// Recursively create a directory and all of its parent components if they
+    /// are missing.
+    pub fn create_dir_all(&mut self, path: impl AsRef<Path>) -> FileSystemResult<()> {
+        let path = path.as_ref();
+
+        let normalized_path = NormalizedPath::new("\\", path)?;
```

Comment:
this is ugly but will be replaced once there is a new and better Path/PAthBuf abstraction